src/compiler/types/generate-event-listener-types.ts

```diff
@@ -0,0 +1,80 @@
+import { dashToPascalCase } from '@utils';
+
+import type * as d from '../../declarations';
+import { updateTypeIdentifierNames } from './stencil-types';
+
+/**
+ * Generates event listener properties for the component's html element type. This is used to allow implementers
+ * to use strict typings when adding and removing event listeners.
+ *
+ * @param cmp The component compiler metadata
+ * @param typeImportData locally/imported/globally used type names, which may be used to prevent naming collisions
+ * @returns additional types information to add event listener method overloads for component's html element type
+ */
+export const generateEventListenerTypes = (
+  cmp: d.ComponentCompilerMeta,
+  typeImportData: d.TypesImportData,
+): { htmlElementEventMap: string[]; htmlElementEventListenerProperties: string[] } => {
+  const tagName = cmp.tagName.toLowerCase();
+  const tagNameAsPascal = dashToPascalCase(tagName);
+  const htmlElementName = `HTML${tagNameAsPascal}Element`;
+  const cmpEventInterface = `${tagNameAsPascal}CustomEvent`;
+  const htmlElementEventMapName = `${htmlElementName}EventMap`;
+  const cmpEvents = cmp.events.filter((cmpEvent) => cmpEvent.complexType.original);
+  if (!cmpEvents.length) {
+    return { htmlElementEventMap: [], htmlElementEventListenerProperties: [] };
+  }
+  return {
+    htmlElementEventMap: getHtmlElementEventMap(cmpEvents, typeImportData, cmp.sourceFilePath, htmlElementEventMapName),
+    htmlElementEventListenerProperties: [
+      `                addEventListener<K extends keyof ${htmlElementEventMapName}>(type: K, listener: (this: ${htmlElementName}, ev: ${cmpEventInterface}<${htmlElementEventMapName}[K]>) => any, options?: boolean | AddEventListenerOptions): void;`,
+      `                addEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions): void;`,
+      `                removeEventListener<K extends keyof ${htmlElementEventMapName}>(type: K, listener: (this: ${htmlElementName}, ev: ${cmpEventInterface}<${htmlElementEventMapName}[K]>) => any, options?: boolean | EventListenerOptions): void;`,
+      `                removeEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | EventListenerOptions): void;`,
```

Comment:
Similar to my comment above, can we add:
```suggestion
      `                removeEventListener<K extends keyof HTMLElementEventMap>(type: K, listener: (this: HTMLElement, ev: HTMLElementEventMap[K]) => any, options?: boolean | EventListenerOptions): void;`,
      `                removeEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | EventListenerOptions): void;`,
```

Reply:
Oh, actually, after re-reviewing the change suggestion, your suggestion is to add

```ts
removeEventListener<K extends keyof HTMLElementEventMap>(type: K, listener: (this: HTMLElement, ev: HTMLElementEventMap[K]) => any, options?: boolean | EventListenerOptions): void;
```

I thought this property and its pairing:

```ts
addEventListener<K extends keyof HTMLElementEventMap>(type: K, listener: (this: HTMLElement, ev: HTMLElementEventMap[K]) => any, options?: boolean | AddEventListenerOptions): void;
```

wouldn't be needed since they would be inherited from the `HTMLElement` interface: https://github.com/microsoft/TypeScript/blob/main/src/lib/dom.generated.d.ts#L10082

I don't mind adding them in since they shouldn't hurt anything, but what were the issues found without these properties?

Reply:
>I don't mind adding them in since they shouldn't hurt anything

Thanks! Changes look good to me!

> but what were the issues found without these properties?

I ran into a few type errors for Ionic Framework components that use `addEventListener` and `removeEventListener` for `KeyboardEvent`s.

After checking out this branch, I reverted the changes in `c3df57ebe0132073663299425e9ae6188df614a0` (as the changes in it resolve the issue), and built it locally by running the following:
```bash
npm run clean \
&& npm ci \
&& npm run build \
&& npm pack
```

I then checked out the Ionic Framework's `main` branch (which, as of this writing, is at `068d0038602485085face4a443e473bf1d5cc227`).
From there, I was able to reproduce this by running the following, starting in the root of the Ionic Framework repo:
```bash
cd core \
&& npm i [PATH_TO_STENCIL_TARBALL] \
&& npm run build \
&& npm run build # second build is intentional
```

After that second build, we get the following errors:
```console
[20:45.6]  @stencil/core
[20:45.7]  [LOCAL DEV] v4.5.0-dev.1697724915.aee7c9e 🌇
[20:47.5]  build, ionic, prod mode, started ...
[20:47.5]  transpile started ...
[20:51.3]  transpile finished in 3.82 s

[ ERROR ]  TypeScript: src/components/picker-internal/picker-internal.tsx:268:7
           No overload matches this call.Overload 1 of 2, '(type: "ionInputModeChange", listener: (this:
           HTMLIonPickerInternalElement, ev: IonPickerInternalCustomEvent<PickerInternalChangeEventDetail>) => any,
           options?: boolean | ... 1 more ... | undefined): void', gave the following error.Argument of type
           '"keypress"' is not assignable to parameter of type '"ionInputModeChange"'.Overload 2 of 2, '(type: string,
           listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions | undefined):
           void', gave the following error.Argument of type '(ev: KeyboardEvent) => void' is not assignable to
           parameter of type 'EventListenerOrEventListenerObject'.Type '(ev: KeyboardEvent) => void' is not assignable
           to type 'EventListener'.Types of parameters 'ev' and 'evt' are incompatible.Type 'Event' is missing the
           following properties from type 'KeyboardEvent': altKey, charCode, code, ctrlKey, and 17 more.

    L267:  } else {
    L268:    el.addEventListener('keypress', this.onKeyPress);
    L269:    this.destroyKeypressListener = () => {

[ ERROR ]  TypeScript: src/components/picker-internal/picker-internal.tsx:270:9
           No overload matches this call.Overload 1 of 2, '(type: "ionInputModeChange", listener: (this:
           HTMLIonPickerInternalElement, ev: IonPickerInternalCustomEvent<PickerInternalChangeEventDetail>) => any,
           options?: boolean | ... 1 more ... | undefined): void', gave the following error.Argument of type
           '"keypress"' is not assignable to parameter of type '"ionInputModeChange"'.Overload 2 of 2, '(type: string,
           listener: EventListenerOrEventListenerObject, options?: boolean | EventListenerOptions | undefined): void',
           gave the following error.Argument of type '(ev: KeyboardEvent) => void' is not assignable to parameter of
           type 'EventListenerOrEventListenerObject'.

    L269:  this.destroyKeypressListener = () => {
    L270:    el.removeEventListener('keypress', this.onKeyPress);
    L271:  };

[ ERROR ]  TypeScript: src/components/popover/utils.ts:441:3
           No overload matches this call.Overload 1 of 2, '(type: keyof HTMLIonPopoverElementEventMap, listener: (this:
           HTMLIonPopoverElement, ev: IonPopoverCustomEvent<void | OverlayEventDetail<any>>) => any, options?: boolean
           | ... 1 more ... | undefined): void', gave the following error.Argument of type '"keydown"' is not
           assignable to parameter of type 'keyof HTMLIonPopoverElementEventMap'.Overload 2 of 2, '(type: string,
           listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions | undefined):
           void', gave the following error.Argument of type '(ev: KeyboardEvent) => Promise<void>' is not assignable to
           parameter of type 'EventListenerOrEventListenerObject'.Type '(ev: KeyboardEvent) => Promise<void>' is not
           assignable to type 'EventListener'.Types of parameters 'ev' and 'evt' are incompatible.Type 'Event' is not
           assignable to type 'KeyboardEvent'.

    L441:    popoverEl.addEventListener('keydown', callback);
    L442:    return () => popoverEl.removeEventListener('keydown', callback);

[ ERROR ]  TypeScript: src/components/popover/utils.ts:442:16
           No overload matches this call.Overload 1 of 2, '(type: keyof HTMLIonPopoverElementEventMap, listener: (this:
           HTMLIonPopoverElement, ev: IonPopoverCustomEvent<void | OverlayEventDetail<any>>) => any, options?: boolean
           | ... 1 more ... | undefined): void', gave the following error.Argument of type '"keydown"' is not
           assignable to parameter of type 'keyof HTMLIonPopoverElementEventMap'.Overload 2 of 2, '(type: string,
           listener: EventListenerOrEventListenerObject, options?: boolean | EventListenerOptions | undefined): void',
           gave the following error.Argument of type '(ev: KeyboardEvent) => Promise<void>' is not assignable to
           parameter of type 'EventListenerOrEventListenerObject'.

    L441:    popoverEl.addEventListener('keydown', callback);
    L442:    return () => popoverEl.removeEventListener('keydown', callback);
    L443:  };

[20:51.3]  build failed in 3.83 s
```

We can group those errors into two groups of two, since two occur in `ion-popover` and two occur in `ion-picker`. I _believe_ they're running into the same issue/limitation of Stencil's generated types (or TypeScript itself?)

Looking at `ion-picker`:
```ts
export const configureKeyboardInteraction = (popoverEl: HTMLIonPopoverElement) => {
	const callback = async (ev: KeyboardEvent): Promise<void> => {
		// implementation elided
	}

	popoverEl.addEventListener('keydown', callback);   // first error
	return () => popoverEl.removeEventListener('keydown', callback); // second error
}
```

You're correct in that having a generated interface like:
```ts
interface HTMLIonPopoverElement extends Components.IonPopover, HTMLStencilElement {  
    addEventListener<K extends keyof HTMLIonPopoverElementEventMap>(type: K, listener: (this: HTMLIonPopoverElement, ev: IonPopoverCustomEvent<HTMLIonPopoverElementEventMap[K]>) => any, options?: boolean | AddEventListenerOptions): void;  
    addEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions): void;  
    removeEventListener<K extends keyof HTMLIonPopoverElementEventMap>(type: K, listener: (this: HTMLIonPopoverElement, ev: IonPopoverCustomEvent<HTMLIonPopoverElementEventMap[K]>) => any, options?: boolean | EventListenerOptions): void;  
    removeEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | EventListenerOptions): void;  
}
```
where `HTMLIonPopoverElement` extends `HTMLStencilElement` (which includes my suggested `addEventListener` and `removeEventListener`) *should* pick up the interface definitions from `HTMLStencilElement`. For some reason, that doesn't appear to be the case though. Prior to applying this change (in Stencil v4.5.0), `addEventListener` and `removeEventListener` _do_ in fact resolve to the typings defined on `HTMLStencilElement` . So something's amuck here, but I think adding the interfaces here are fine for our purposes.

Reply:
That is odd, but yeah, adding the additional overloads would do more help than harm, so I think it'd be fair just to add them. I'll add the additional event listeners to the PR and update everything later tonight.